### PR TITLE
Use 'ShortText' as internal representation of `PackageName`

### PR DESCRIPTION
**Pull request by hvr**

### CODE REVIEW

File: Cabal/Distribution/ShortText.hs [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+
+----------------------------------------------------------------------------
+-- UTF8 helpers
+
+-- | Decode 'String' from UTF8-encoded octets.
+--
+-- Invalid data will be decoded as the replacement character (@U+FFFD@)
+--
+-- See also 'encodeStringUtf8'
+decodeStringUtf8 :: [Word8] -> String
```
Comment (phadej) [inline Cabal/Distribution/ShortText.hs]: could this leave somewhere in `.Utils.` or `Compat``? I'll need exactly the same functions in `parsec-parser` too

Comment (hvr) [inline Cabal/Distribution/ShortText.hs]: @phadej I had to move it out of `.Simple.Utils` due to cyclic deps; I'm open to a new home for it... please suggest a concrete alternative place for those

Comment (phadej) [inline Cabal/Distribution/ShortText.hs]: as I said on IRC, somewhere out of the `D.Simple`, as parsec is outside too.

**Comment by 23Skidoo** [comment]
Looks good, let's merge once it's green!

**Comment by hvr** [comment]
Just for the record: `cabal info zzz +RTS -s` before this patch

```
cabal.0: There is no package named 'zzz'.
You may need to run 'cabal update' to get the latest list of available
packages.
     958,149,400 bytes allocated in the heap
     486,587,248 bytes copied during GC
      44,009,784 bytes maximum residency (19 sample(s))
         959,168 bytes maximum slop
             122 MB total memory in use (0 MB lost due to fragmentation)

                                     Tot time (elapsed)  Avg pause  Max pause
  Gen  0      1802 colls,     0 par    0.290s   0.290s     0.0002s    0.0009s
  Gen  1        19 colls,     0 par    0.231s   0.231s     0.0122s    0.0518s

  TASKS: 4 (1 bound, 3 peak workers (3 total), using -N1)

  SPARKS: 0 (0 converted, 0 overflowed, 0 dud, 0 GC'd, 0 fizzled)

  INIT    time    0.001s  (  0.001s elapsed)
  MUT     time    0.516s  (  0.812s elapsed)
  GC      time    0.521s  (  0.521s elapsed)
  EXIT    time    0.004s  (  0.004s elapsed)
  Total   time    1.044s  (  1.338s elapsed)

  Alloc rate    1,855,307,269 bytes per MUT second

  Productivity  50.0% of total user, 39.0% of total elapsed

gc_alloc_block_sync: 0
whitehole_spin: 0
gen[0].sync: 0
gen[1].sync: 0
```

after this patch:

```
cabal.1: There is no package named 'zzz'.
You may need to run 'cabal update' to get the latest list of available
packages.
     858,495,280 bytes allocated in the heap
     455,890,816 bytes copied during GC
      41,306,624 bytes maximum residency (19 sample(s))
       1,427,064 bytes maximum slop
             105 MB total memory in use (0 MB lost due to fragmentation)

                                     Tot time (elapsed)  Avg pause  Max pause
  Gen  0      1612 colls,     0 par    0.257s   0.256s     0.0002s    0.0006s
  Gen  1        19 colls,     0 par    0.231s   0.231s     0.0121s    0.0546s

  TASKS: 4 (1 bound, 3 peak workers (3 total), using -N1)

  SPARKS: 0 (0 converted, 0 overflowed, 0 dud, 0 GC'd, 0 fizzled)

  INIT    time    0.001s  (  0.001s elapsed)
  MUT     time    0.427s  (  0.781s elapsed)
  GC      time    0.487s  (  0.487s elapsed)
  EXIT    time    0.004s  (  0.004s elapsed)
  Total   time    0.921s  (  1.273s elapsed)

  Alloc rate    2,008,686,163 bytes per MUT second

  Productivity  47.0% of total user, 34.0% of total elapsed

gc_alloc_block_sync: 0
whitehole_spin: 0
gen[0].sync: 0
gen[1].sync: 0
```

**Pull request merged**